zebra-network/src/peer/connection.rs

```diff
@@ -506,123 +509,143 @@ where
 
         // XXX(hdevalence) this is truly horrible, but let's fix it later
 
-        // Inner match returns Result with the new state or an error.
-        // Outer match updates state or fails.
+        // Inner matches return a Result with a new state or an (error, Option<oneshot::Sender>)
+        // Middle match returns Result with the new state or the (error, Option<oneshot::Sender>)
+        // Outer match updates state or fails, and sends the error on the Sender if it is Some
```

Comment:
Now we have 3 layers of matches, which is extra horrible, so we should fix it now 😱

Reply:
Fixed in 6ed420ee238a031eb57b152e816663e0c98880ff